network/tcp_client.cc

```diff
@@ -0,0 +1,490 @@
+                                         receive_or.status().message()));
+    }
+
+    auto response = std::move(receive_or.value());
+    if (response->GetMessageType() != static_cast<uint32_t>(MessageType::HANDSHAKE_RESPONSE))
+    {
+        return absl::FailedPreconditionError(
+        absl::StrCat("PerformHandshake: Unexpected message type in Handshake response "
+                     "(Expected: ",
+                     MessageType::HANDSHAKE_RESPONSE,
+                     ", Got: ",
+                     response->GetMessageType(),
+                     ")."));
+    }
+
+    auto* hs_response = dynamic_cast<HandShakeResponse*>(response.get());
+    if (!hs_response)
+    {
+        return absl::InternalError(
+        "PerformHandshake: Failed to cast received message to HandShakeResponse.");
+    }
+
+    std::cout << "Client: Server Handshake (Server v" << hs_response->GetMajorVersion() << "."
+              << hs_response->GetMinorVersion() << ")" << std::endl;
+
+    if (hs_response->GetMajorVersion() != hs_request.GetMajorVersion() ||
+        hs_response->GetMinorVersion() != hs_request.GetMinorVersion())
+    {
+        return absl::FailedPreconditionError(
+        absl::StrCat("PerformHandshake: Handshake version mismatch. Server is v",
+                     hs_response->GetMajorVersion(),
+                     ".",
+                     hs_response->GetMinorVersion(),
+                     " Client requires v",
+                     hs_request.GetMajorVersion(),
+                     ".",
+                     hs_request.GetMinorVersion()));
+    }
+    std::cout << "Client: Handshake versions compatible." << std::endl;
+    return absl::OkStatus();
+}
+
+absl::Status TcpClient::StartKeepAlive()
+{
+    if (m_keep_alive.running.load())
+    {
+        return absl::AlreadyExistsError("StartKeepAlive: Keep-alive is already running.");
+    }
+    if (!IsConnected())
+    {
+        return absl::FailedPreconditionError("StartKeepAlive: Client is not connected.");
+    }
+
+    m_keep_alive.running.store(true);
+    m_keep_alive.thread = std::thread(&TcpClient::KeepAliveLoop, this);
+
+    std::cout << "Client: Keep-alive thread started. Interval: " << m_keep_alive.interval_sec
+              << " s, Ping Timeout: " << m_keep_alive.ping_timeout_ms << " ms." << std::endl;
+    return absl::OkStatus();
+}
+
+void TcpClient::KeepAliveLoop()
+{
+    while (m_keep_alive.running.load())
+    {
+        std::unique_lock<std::mutex> lk(m_keep_alive.mutex);
+        if (m_keep_alive.cv.wait_for(lk, std::chrono::seconds(m_keep_alive.interval_sec), [this] {
+                return !m_keep_alive.running.load();
+            }))
+        {
+            std::cout << "KeepAliveLoop: Client Keep-Alive Thread: Stop signaled via CV."
+                      << std::endl;
+            break;
+        }
+        lk.unlock();
+
+        if (!m_keep_alive.running.load())
```

Comment:
isn't the `wait_for` itself returning true if !m_keep_alive.running.load() was the reason it woke up? do you still need this extra check here?